src/glue/jobs/json_s3_to_parquet.py

```diff
@@ -132,105 +141,246 @@ def add_index_to_table(table_key, table_name, processed_tables, unprocessed_tabl
+    """Function to convert JSON schema to specs parameter in
+    DynamicFrame.resolveChoice format as a list of tuples
+
+    Args:
+        json_schema (dict): The json schema to convert
+        path (list, optional): Path used in recursion in function.
+            Defaults to [].
+
+    Returns:
+        list: list of tuples in the expected form:
+         [
+             (path_to_field1 : cast:<field1_type>),
+             (path_to_field2 : cast:<field2_type>)
+         ]
+    """
+    # handles the condition in which the top level json schema is like "testField":[...]
+    if isinstance(json_schema, list):
+        json_schema = json_schema[0]
+    specs = []
+    for field, field_type in json_schema.items():
+        # handles the condition in which the nested json schema is like {"testField": ...}
+        if isinstance(field_type, dict):
+            inner_specs = convert_json_schema_to_specs(field_type, path + [field])
+            specs.extend(inner_specs)
+
+        # handles the condition in which the nested json schema is like "testField":[...]
+        elif isinstance(field_type, list):
+            # handles the condition in which the nested json schema is like "testField":[{...}]
+            if isinstance(field_type[0], dict):
+                inner_specs = convert_json_schema_to_specs(
+                    field_type, path + [f"{field}[]"]
+                )
+                specs.extend(inner_specs)
+            # Separates out the condition of a json schema scenario like "testField": ["string"]
+            # testField:array<string> (hive schema version).
+            # We currently don't have plans to cast something like that to another data type
+            else:
+                pass
+        else:
+            # handles the condition in which the nested json schema is like "testField":some_value
+            specs.append((".".join(path + [field]), f"cast:{field_type}"))
+    return specs
+
+
 def main():
+    glue_context = GlueContext(SparkContext.getOrCreate())
+
     # Get args and setup environment
     args, workflow_run_properties = get_args()
-    glue_context = GlueContext(SparkContext.getOrCreate())
     logger = glue_context.get_logger()
     job = Job(glue_context)
     job.init(args["JOB_NAME"], args)
 
     # Get table info
     table_name = args["table"]
     table = get_table(
-            table_name=table_name,
-            database_name=workflow_run_properties["database"],
-            glue_context=glue_context
+        table_name=table_name,
+        database_name=workflow_run_properties["database"],
+        glue_context=glue_context,
     )
+
+    # cast table's glue types to crawlers' json schema types
+    if args["cast_glue_data_types"]:
+        table = cast_glue_data_types(
+            dynamic_frame=table,
+            glue_database_name=table_name,
+            glue_table_name=workflow_run_properties["database"],
+        )
```

Comment:
You should be replacing L81-83 with the `cast_glue_data_types` logic. Otherwise we've already resolved the types (incorrectly) in `get_table` above.

Reply:
Ah that was what I wasn't sure about it in the end, because I thought the `get_table`'s resolveChoice wouldn't permanently cast it